node/network/approval-distribution/src/lib.rs

```diff
@@ -537,67 +768,109 @@ impl State {
+					*required_routing
 				}
 			},
+			&self.peer_views,
 		)
 		.await;
 	}
 
+	async fn process_incoming_assignments<Context, R>(
+		&mut self,
+		ctx: &mut Context,
+		metrics: &Metrics,
+		peer_id: PeerId,
+		assignments: Vec<(IndirectAssignmentCertV2, CandidateBitfield)>,
+		rng: &mut R,
+	) where
+		R: CryptoRng + Rng,
+	{
+		for (assignment, claimed_indices) in assignments {
+			if let Some(pending) = self.pending_known.get_mut(&assignment.block_hash) {
+				let block_hash = &assignment.block_hash;
+				let validator_index = assignment.validator;
+
+				gum::trace!(
+					target: LOG_TARGET,
+					%peer_id,
+					?block_hash,
+					?claimed_indices,
+					?validator_index,
+					"Pending assignment",
+				);
+
+				pending.push((peer_id, PendingMessage::Assignment(assignment, claimed_indices)));
+
+				continue
+			}
+
+			self.import_and_circulate_assignment(
+				ctx,
+				metrics,
+				MessageSource::Peer(peer_id),
+				assignment,
+				claimed_indices,
+				rng,
+			)
+			.await;
+		}
+	}
+
 	async fn process_incoming_peer_message<Context, R>(
 		&mut self,
 		ctx: &mut Context,
 		metrics: &Metrics,
 		peer_id: PeerId,
-		msg: protocol_v1::ApprovalDistributionMessage,
+		msg: Versioned<
+			protocol_v1::ApprovalDistributionMessage,
+			protocol_vstaging::ApprovalDistributionMessage,
+		>,
 		rng: &mut R,
 	) where
 		R: CryptoRng + Rng,
 	{
 		match msg {
-			protocol_v1::ApprovalDistributionMessage::Assignments(assignments) => {
+			Versioned::VStaging(protocol_vstaging::ApprovalDistributionMessage::Assignments(
+				assignments,
+			)) => {
+				gum::trace!(
+					target: LOG_TARGET,
+					peer_id = %peer_id,
+					num = assignments.len(),
+					"Processing assignments from a peer",
+				);
+				self.process_incoming_assignments(ctx, metrics, peer_id, assignments, rng).await;
+			},
+			Versioned::V1(protocol_v1::ApprovalDistributionMessage::Assignments(assignments)) => {
 				gum::trace!(
 					target: LOG_TARGET,
 					peer_id = %peer_id,
 					num = assignments.len(),
 					"Processing assignments from a peer",
 				);
-				for (assignment, claimed_index) in assignments.into_iter() {
-					if let Some(pending) = self.pending_known.get_mut(&assignment.block_hash) {
-						let message_subject = MessageSubject(
-							assignment.block_hash,
-							claimed_index,
-							assignment.validator,
-						);
-
-						gum::trace!(
-							target: LOG_TARGET,
-							%peer_id,
-							?message_subject,
-							"Pending assignment",
-						);
-
-						pending
-							.push((peer_id, PendingMessage::Assignment(assignment, claimed_index)));
-
-						continue
-					}
 
-					self.import_and_circulate_assignment(
-						ctx,
-						metrics,
-						MessageSource::Peer(peer_id),
-						assignment,
-						claimed_index,
-						rng,
-					)
-					.await;
-				}
+				self.process_incoming_assignments(
+					ctx,
+					metrics,
+					peer_id,
+					assignments
+						.into_iter()
+						.map(|(cert, candidate)| (cert.into(), candidate.into()))
```

Comment:
doesn't look like we sanity-check the `CandidateIndex` is in a reasonable range before converting it to a bitfield.

I believe a malicious node could send us `(cert_v1, candidate_index(u32::max))` and force us to allocate.

It may be better for us to not eagerly convert to `CandidateBitfield` in the approval entry, and instead use an `enum`.

Reply:
yeah, good point!